### PR TITLE
Feature/metric point generator

### PR DESCRIPTION
**Pull request by faneshion**
#301 #307   revised datapack and point_generator.

### CODE REVIEW

File: matchzoo/engine/base_preprocessor.py [hunk 1]
```diff
@@ -57,28 +58,28 @@ def fit_transform(self, inputs: list, stage: str) -> datapack.DataPack:
 
     def _make_output(
         self,
-        output: list,
-        mapping: pd.DataFrame,
+        relation: typing.Union[list, np.ndarray],
```
Comment (bwanglzu) [inline matchzoo/engine/base_preprocessor.py]: relation is pandas?

File: matchzoo/datapack.py [hunk 2]
```diff
@@ -13,58 +13,61 @@ class DataPack(object):
...
 
-        :param data: Input data, could be list-like objects
-            or :class:`numpy.ndarray`.
-        :param mapping: Store the mapping between left document
+        :param relation: Store the relation between left document
             and right document use ids.
+        :param content: Store the content of ids.
         :param context: Hyper-parameter fitted during
             pre-processing stage.
         :param columns: List of column names of the :attr:`data`
```
Comment (bwanglzu) [inline matchzoo/datapack.py]: now we do not have `data` column

File: matchzoo/engine/base_generator.py [hunk 3]
```diff
@@ -55,39 +58,34 @@ def __getitem__(self, idx: int) -> typing.Tuple[dict, typing.Any]:
...
         return (self.num_instances + self.batch_size - 1) // self.batch_size
 
+    def __iter__(self) -> typing.Tuple[dict, list]:
+        """Create an generator."""
+        if self.stage == 'train':
+            while True:
+                for item in (self[i] for i in range(len(self))):
+                    yield item
+        else:
+            for item in (self[i] for i in range(len(self))):
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]: guess we need a while here as well?

File: matchzoo/engine/base_preprocessor.py (hunk 1 above)
Comment (faneshion) [reply matchzoo/engine/base_preprocessor.py]: Here, the `_make_output` function is to construct a `DataPack` object based on 'relation & content & context & columns'. The construction function of `DataPack` only accept 'list or np.ndarray' as input for 'relation'. Thus, I do not use pandas for `relation`. In fact, the preprocess module transform a datapack (from segmendation) to a new datapack (by _make_output). During the transformation, the relation often keeps the same. Thus, a more elegant way is to re-assign the content rather than re-construct a datapack. What do you think ?

File: matchzoo/datapack.py (hunk 2 above)
Comment (faneshion) [reply matchzoo/datapack.py]: it's done~

File: matchzoo/engine/base_preprocessor.py [hunk 4]
```diff
@@ -57,28 +57,25 @@ def fit_transform(self, inputs: list, stage: str) -> datapack.DataPack:
...
             a :class:`DataPack` object.
         """
-        _, columns_data, _ = self._get_columns(stage=stage)
-        return datapack.DataPack(data=output,
-                                 mapping=mapping,
-                                 context=context,
-                                 columns=columns_data)
+        data.content = content
+        data.context = context
+        return data
```
Comment (bwanglzu) [inline matchzoo/engine/base_preprocessor.py]: i don't understand these three lines.

Comment (faneshion) [reply matchzoo/engine/base_preprocessor.py]: The preprocessor are designed to deal with the contents, while the relations and columns are keep the same. In the segmentation function, it have constructed the DataPack object. All the transformers are utilized to transfer the content. Thus, we can use the transformed content and context to update the DataPack object. Rather than to re-construct a new DataPack object.

Comment (faneshion) [reply matchzoo/engine/base_preprocessor.py]: The preprocessor are designed to deal with the contents, while the relations and columns are keep the same. In the segmentation function, it have already constructed the DataPack object. All the transformers are utilized to transfer the content. Thus, we can use the transformed content and context to update the DataPack object. Rather than to re-construct a new DataPack object.

File: matchzoo/engine/base_preprocessor.py [hunk 5]
```diff
@@ -104,50 +105,53 @@ def segmentation(self, inputs: list, stage: str) -> datapack.DataPack:
         """
         Convert user input into :class:`DataPack` consist of two tables.
 
-        The `data` table stores the id with it's corresponded input text.
-        The `mapping` table stores the mapping between `text_left` and
+        The `content` dictstores the id with it's corresponded input text.
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/engine/base_preprocessor.py]: it's done~

File: matchzoo/preprocessor/dssm_preprocessor.py [hunk 6]
```diff
@@ -95,10 +95,11 @@ def fit(self, inputs: typing.List[tuple]):
         # 1. Used for build vocabulary of tri-letters (get dimension).
         # 2. Cached tri-letters can be further used to perform input
         #    transformation.
-        df = self._datapack.dataframe
+        content = self._datapack.content
 
-        for idx, text in tqdm(zip(df['id'], df['text'])):
+        for idx, text in tqdm(content.items()):
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/preprocessor/dssm_preprocessor.py]: done~
